explorer/interpreter/interpreter.cpp

```diff
@@ -1503,6 +1503,72 @@ auto Interpreter::StepStmt() -> ErrorOr<Success> {
         }
       }
     }
+    case StatementKind::For: {
+      static int CURRENT_INDEX_POS = 1;
+      static int TARGET_VAR_POS = 0;
+      static int END_INDEX_POS = 2;
+      if (act.pos() == 0) {
+        return todo_.Spawn(
+            std::make_unique<ExpressionAction>(&cast<For>(stmt).loop_target()));
+      }
+      if (act.pos() == 1) {
+        Nonnull<const TupleValue*> source_array =
+            cast<const TupleValue>(act.results().back());
+
+        auto end_index = static_cast<int>(source_array->elements().size());
+        auto start_index = 0;
+
+        if (start_index < end_index) {
+          act.AddResult(arena_->New<IntValue>(start_index));
+          act.AddResult(arena_->New<IntValue>(end_index));
+          return todo_.Spawn(std::make_unique<PatternAction>(
+              &cast<For>(stmt).variable_declaration()));
+        }
+        return todo_.FinishAction();
```

Comment:
I think it'd be a little clearer to reverse this:

```suggestion
        if (end_index > 0) {
          return todo_.FinishAction();
        }
        act.AddResult(arena_->New<IntValue>(start_index));
        act.AddResult(arena_->New<IntValue>(end_index));
        return todo_.Spawn(std::make_unique<PatternAction>(
            &cast<For>(stmt).variable_declaration()));      
```

That way the general case is handled in "main line" code, and the corner case is handled in a side branch.